docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,157 @@
+- num_iters - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+For now, we assume that the `Block` classes know how to use these parameters and there is no need to use any sort of templating in the custom flows that could use these parameters.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+For now, we assume that the `LLMBlock` class will choose the appropriate model prompt based on the family and that there is no need to specify a custom prompt.
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+We will continue to use these config files unchanged, and custom files can be specified with an absolute path. Relative paths are assumed to be relative to the Python package e.g. `configs/skills/...`.
+
+### Future - Model Serving
+
+Custom pipelines may have more unique model serving requirements. Instead of serving just one model, we may need to launch the model server with a model and an additional model with adapter.
+
+Right now the sdg code takes two parameters for each block - the OpenAI client instance, and the model ID. This should be sufficient. vLLM, for example, can host both a model and a model+adapter under two different model IDs.
+
+In future, it may make sense to define some model serving config (e.g. which adapter to use for SDG) as part of a flow definition. We will address this in a future if enhancement, if needed.
+
+### CLI Integration
+
+As of the current version of `ilab`, it supports `simple` and `full` as parameters to `--pipeline` to select one of the two types of built-in pipelines included in the library. Once we have support for loading custom pipelines, we need a way for these to be specified with the CLI. Here is a proposed CLI UX for this:
+
+> `ilab data generate`
+
+Use the default pipeline, `simple`.
+
+> `ilab data generate --pipeline full`
+
+Use the built-in `full` pipeline.
+
+> `ilab data generate --pipeline full --pipeline-extend path/to/config.yaml`
+
+Use the `full` pipeline, but extend it with the pipeline defined in the provided config file.
+
+> `ilab data generate --pipeline none --pipeline-extend path/to/other_config.yaml`
```

Comment:
```suggestion
> `ilab data generate --pipeline-extend path/to/other_config.yaml`
```
This seems clunky. Would it be more usable without `pipeline` specified? Therefore, specifying the extended pipeline with no pipeline specified then it is assumed extended only.

Reply:
I agree it's clunky, I don't love it

But I think it's important `--pipeline` has a default value that isn't affected by other arguments - so your suggestion reads "extend the default pipeline" IMO

@russellb

Reply:
I think having a default value for `--pipeline` that changes depending on other values seems less desirable than having to specify `--pipeline none`. This is probably the least likely combination of values to use, anyway.

Another option could be allowing both file paths and the aliases (`simple`, `full`) as arguments to `--pipeline`. So a custom pipeline that isn't extending one could just be:

```
ilab data generate --pipeline path/to/custom_pipeline.yaml
```

Reply:
or yet another alternative, drop `--pipeline-extend` and just allow `--pipeline` to be specified multiple times.

```
ilab data generate --pipeline simple
```

```
ilab data generate --pipeline path/to/custom_pipeline.yaml
```

```
ilab data generate --pipeline full --pipeline path/to/extensions.yaml
```

Reply:
Ok, aybe its best to be explicit with `pipeline`. I am ok to leave it as is for now.